### PR TITLE
feat(auth): add persistent Studio user management

### LINKED ISSUE #2161: feat(auth): add persistent Studio users and session management
## Problem
Studio authentication currently reads users from configuration and retains bearer sessions only in process memory. It does not expose a stable user ID, account disablement, password rotation, or persistent session revocation.

## Scope
- Add Studio-local user and session tables, distinct from RocketMQ ACL users.
- Bootstrap the first database users from existing `studio.auth.users` only when the table is empty.
- Use immutable user IDs, salted password hashes, and persisted token hashes.
- Add basic create/list/enable/disable/reset-password operations for administrators.
- Add self-service password change and revoke sessions on disable/password change.
- Add a minimal Studio user-management UI.

## Out of scope
Roles beyond the existing admin flag, organizations/tenants, SSO/OIDC, and external identity providers.

### PR DESCRIPTION
**Pull request by Aias00**
## Summary
- add `rmq_studio_user` and `rmq_studio_session` schema plus MyBatis entities/mappers
- use immutable user IDs, PBKDF2 password hashes, and SHA-256 token hashes for persisted sessions
- bootstrap initial database users from existing `studio.auth.users` only when no Studio user exists
- add admin user list/create/enable/disable/password-reset APIs and a minimal UI
- add self-service password change; disable and password changes revoke all active sessions
- deliver browser sessions in a `HttpOnly`, `Secure`, `SameSite` cookie; browser JavaScript no longer receives or stores the token
- retain Authorization-header support for non-browser API clients

## Scope
This deliberately excludes roles beyond the existing admin flag, organizations/tenants, and SSO/OIDC.

Closes #2161

## Configuration
- `STUDIO_AUTH_SESSION_COOKIE_NAME` defaults to `rmq_studio_session`.
- Production defaults to `Secure=true` and `SameSite=Strict`.
- The `dev` profile automatically uses `session-cookie-secure=false` so local HTTP works without environment configuration.
- When the frontend uses a different origin, configure `STUDIO_CORS_ALLOWED_ORIGINS` and keep credentialed requests enabled.

## Verification
- `JAVA_HOME=$( /usr/libexec/java_home -v 21 ) mvn -Dtest=AuthCookieTest,AuthCookieProfileTest,AuthControllerTest,AuthInterceptorTest,AuthServiceTest,AuthServiceDatabaseTest,PasswordHasherTest test` (64 tests)
- `npm run lint` (one pre-existing `MainLayout` hooks warning remains)
- `npm test -- --run src/App.test.tsx src/layouts/MainLayout.test.tsx src/api/client.test.ts src/api/auth.test.ts src/api/ai.test.ts src/stores/authStorage.test.ts src/pages/studio/__tests__/Ops.test.tsx` (48 tests)
- `npm run build`

### CODE REVIEW

**Review by RockteMQ-AI — APPROVED** [review]
### Summary
Persistent Studio user management — comprehensive authentication overhaul with solid security practices.

This PR adds database-backed user management to RocketMQ Studio, replacing the in-memory configuration-based authentication. The implementation follows security best practices and includes proper migration strategy.

### Findings

**Security Implementation — Excellent:**
- **[Info]** PasswordHasher.java — PBKDF2WithHmacSHA256 with 210,000 iterations, 16-byte salt, 256-bit key. Industry standard for password hashing.
- **[Info]** Session tokens — SHA-256 hashed before database storage, preventing token leakage if database is compromised
- **[Info]** AuthCookie.java — HttpOnly, Secure, SameSite=Strict cookies. Proper cookie security flags
- **[Info]** Token generation — Uses SecureRandom with 32-byte entropy, sufficient for session tokens

**Architecture — Well-designed:**
- **[Info]** Migration strategy — Bootstrap from config users to database only when no Studio users exist. Clean transition path
- **[Info]** Session management — In-memory cache (ConcurrentHashMap) for active sessions, database for persistence. Good balance of performance and durability
- **[Info]** Session revocation — Password changes and user disable revoke all active sessions. Proper security hygiene
- **[Info]** Expired session cleanup — Scheduled purge prevents database bloat

**Test Coverage — Comprehensive:**
- **[Info]** 5 test files covering AuthController, AuthCookie, AuthService, and PasswordHasher
- **[Info]** Tests verify both happy paths and security edge cases

### Suggestions

**Minor:**
- **[Info]** PasswordHasher.java:24 — Consider increasing iterations to 600,000 (OWASP 2023 recommendation for SHA-256). Current 210,000 is acceptable but could be stronger for new deployments
- **[Info]** Consider adding rate limiting to login endpoint to prevent brute-force attacks (separate PR)

**Documentation:**
- **[Info]** Consider documenting the migration path in README or migration guide for existing deployments

### Cross-repo Note
This is a Studio-specific change. No impact on core RocketMQ or client SDKs.

### Overall
**Approve** — This is a well-executed feature PR with strong security foundations, proper session management, and comprehensive tests. The migration strategy is clean and backward-compatible.

---
*Automated review by @RockteMQ-AI*

**Review by lizhimins — CHANGES_REQUESTED** [review]
Solid design overall — the PBKDF2 parameters, constant-time comparison, session revocation, last-admin protection and CORS tightening are all in good shape. Please address the following before merge:

1. **Blocking: token distribution contradiction.** `LoginVO.token` is annotated `@JsonIgnore`, so non-browser clients can never obtain a bearer token, while the `AuthCookie` comments say the header channel is kept for exactly that case. Decide and implement one coherent story (e.g. keep the token in the JSON body for API clients, or document and test the cookie-only flow).
2. **First-login bootstrap race:** concurrent first logins can hit the unique-key constraint and one side gets a raw 500. Please add a lock or conflict-tolerant upsert.
3. Minor: writing `last_seen_at` on every request is noisy — consider throttling. PBKDF2 iterations are below the current OWASP recommendation (600k) — acceptable now, follow-up welcome.

Looking forward to the fixed version!

**Comment by Aias00** [comment]
Addressed the review findings in a536448e:

- Browser login remains cookie-only; non-browser clients can explicitly request a JSON bearer token with `X-RocketMQ-Studio-Session-Delivery: bearer`. The API contract and tests cover both delivery modes.
- Bootstrap user creation now tolerates duplicate-key races between Studio instances.
- `last_seen_at` writes are throttled to once per five minutes per session.

Validation: `JAVA_HOME=$( /usr/libexec/java_home -v 21 ) mvn -Dtest=AuthCookieTest,AuthControllerTest,AuthServiceDatabaseTest,AuthServiceTest,LoginVOTest test` (33 tests passed).

PBKDF2 work-factor tuning and login rate limiting remain intentionally out of this focused review update.

**Comment by Aias00** [comment]
Follow-up update:

- `GET /api/auth/status` now returns the authenticated user identity when a session is valid.
- The frontend refreshes its display identity from that server-authoritative response instead of relying only on local storage.
- Legacy configured users may have no persistent `userId`; the frontend now represents that explicitly.

Validated with Java 21 `mvn -Dtest=AuthControllerTest test` (9 tests), frontend targeted tests (15 tests), and `npm run build`.

**Review by RockteMQ-AI — COMMENTED** [review]
### Summary
This PR adds persistent user management to RocketMQ Studio with database-backed authentication, replacing the in-memory session store. The implementation includes PBKDF2 password hashing, session management, admin user controls, and a UI for user administration.

**Overall: Security-conscious implementation with proper cryptographic practices. This is a large feature addition (47 files) that requires careful review of the authentication flow.**

### Security Review (High-Level)

**✅ Strengths:**
- **Password hashing**: PBKDF2WithHmacSHA256 with 210k iterations, proper salt, constant-time comparison
- **Token storage**: Sessions store SHA-256 hashed tokens, not plaintext
- **Session management**: Proper expiration, cleanup, and revocation on password change
- **Admin protection**: Cannot disable the last enabled administrator
- **Error messages**: Generic "Invalid username or password" prevents username enumeration

**⚠️ Areas for Review:**
1. **Database migration**: The PR adds new schema (`rmq_studio_user`, `rmq_studio_session`). Ensure migration scripts handle upgrades from previous versions gracefully
2. **Bootstrap users**: The `ensureBootstrapUsers()` method creates initial users from config. Verify this only runs once and doesn't overwrite existing users
3. **Session cleanup**: The `@Scheduled` cleanup runs every 5 minutes. Consider making this configurable for high-traffic deployments
4. **Password validation**: The `validatePassword()` method should enforce minimum length and complexity requirements
5. **Rate limiting**: Consider adding rate limiting to login endpoints to prevent brute-force attacks

### Findings
- **[Info]** PasswordHasher.java:28 — Good cryptographic practices
- **[Warning]** AuthService.java:180 — Ensure login failures don't log usernames
- **[Info]** AuthService.java:220 — Proper token hashing and session revocation

### Suggestions
1. Add integration tests for the bootstrap user creation flow
2. Document the migration path from config-based users to database users
3. Consider adding audit logging for admin actions (user create/disable/password reset)
4. Add password complexity validation (min length, no common passwords)
5. Consider adding account lockout after N failed login attempts

### Testing
The PR includes comprehensive test coverage for authentication flows. Verify that:
- Login with wrong credentials returns generic error
- Disabled users cannot login
- Sessions expire correctly
- Password changes invalidate all active sessions
- Last admin cannot be disabled

---
*Automated review by github-manager*

File: server/src/main/java/org/apache/rocketmq/studio/auth/PasswordHasher.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.Base64;
+
+@Component
+public class PasswordHasher {
+
+    private static final String ALGORITHM = "PBKDF2WithHmacSHA256";
+    private static final int ITERATIONS = 210_000;
+    private static final int KEY_LENGTH = 256;
+    private static final int SALT_LENGTH = 16;
+    private static final int MIN_ITERATIONS = 100_000;
```
Comment (RockteMQ-AI) [inline server/src/main/java/org/apache/rocketmq/studio/auth/PasswordHasher.java:28]: **[Info]** Good security practice: PBKDF2WithHmacSHA256 with 210k iterations, proper salt generation, and constant-time comparison via `MessageDigest.isEqual()`. The iteration bounds check (100k-1M) prevents downgrade attacks.

File: server/src/main/java/org/apache/rocketmq/studio/auth/AuthService.java [hunk 2]
```diff
@@ -17,153 +17,393 @@
...
 
-        LoginVO.UserInfo user = authenticate(request);
-        long now = clock.millis();
-        purgeExpiredSessions(now);
-        int tokenTtlSeconds = sessionTimeoutSeconds();
-        String token = "studio-jwt-" + UUID.randomUUID();
-        activeTokens.put(token, new AuthSession(user, now + tokenTtlSeconds * 1000L));
+    public void changePassword(String userId, String currentPassword, String newPassword,
+                               boolean requireCurrentPassword) {
+        requireDatabaseBacked();
```
Comment (RockteMQ-AI) [inline server/src/main/java/org/apache/rocketmq/studio/auth/AuthService.java:180]: **[Warning]** The `loginDatabaseUser` method throws generic "Invalid username or password" for both wrong username and wrong password, which is good for security. However, ensure this error message is not logged with the username to avoid username enumeration via logs.

File: server/src/main/java/org/apache/rocketmq/studio/auth/AuthService.java [hunk 3]
```diff
@@ -17,153 +17,393 @@
...
+        }
+        int tokenTtlSeconds = sessionTimeoutSeconds();
+        String token = newBearerToken();
+        LocalDateTime current = now();
+        RmqStudioSession session = new RmqStudioSession();
+        session.setId(UUID.randomUUID().toString());
+        session.setUserId(user.getId());
+        session.setTokenHash(tokenHash(token));
+        session.setCreatedAt(current);
+        session.setLastSeenAt(current);
```
Comment (RockteMQ-AI) [inline server/src/main/java/org/apache/rocketmq/studio/auth/AuthService.java:220]: **[Info]** Good: Sessions are stored with hashed tokens (SHA-256) rather than plaintext, and sessions are properly revoked on password change and user disable.

**Comment by Aias00** [comment]
Addressed in the latest branch commits:

- Browser logins use the HttpOnly session cookie; API clients explicitly request a bearer token with `X-RocketMQ-Studio-Session-Delivery: bearer`, covered by `AuthControllerTest`.
- Bootstrap user inserts tolerate concurrent duplicate-key creation, covered by `AuthServiceDatabaseTest`.
- `last_seen_at` writes are throttled by `LAST_SEEN_UPDATE_INTERVAL`, with fresh and stale-session coverage.

Validated with `JAVA_HOME=$(/usr/libexec/java_home -v 21) mvn -Dtest=AuthControllerTest,AuthServiceDatabaseTest,AuthCookieTest test`.

**Comment by Aias00** [comment]
Both blocking items are addressed in the current branch. Commit `09423b24` defines an explicit login delivery contract: browsers receive an HttpOnly cookie and API clients request a bearer token with `X-RocketMQ-Studio-Session-Delivery: bearer`; controller tests cover both behavior. The same commit catches the bootstrap `DuplicateKeyException` and adds database regression coverage. Re-ran `AuthControllerTest`, `AuthServiceDatabaseTest`, `AuthCookieTest`, and `AuthCookieProfileTest` successfully.

**Comment by Aias00** [comment]
Superseded by #2313, which consolidates this change and is CLEAN with all CI checks passing.
